torch/functional.py

```diff
@@ -599,3 +600,111 @@ def argsort(input, dim=None, descending=False):
+            =====  ============================  ==========================
+        dim ({int, 2-tuple of ints, 2-list of ints}, optional): If it is a int,
+        vector norm will be calculated, if it is 2-tuple of ints, matrix norm
+        will be calculated. If the value is None, matrix norm will be calculated
+        when the input tensor only has two dimensions, vector norm will be
+        calculated when the input tensor only has one dimension. If the input
+        tensor has more than two dimensions, the vector norm will be applied to
+        last dimension.
+        keepdim (bool): whether the output tensors have :attr:`dim`
+            retained or not. Ignored if ``dim=None and out=None``.
+        out (Tensor, optional) – the output tensor. Ignored if
+        ``dim=None and out=None``.
+
+    Example::
+        >>> import torch
+        >>> a = torch.arange(9, dtype= torch.float) - 4
+        >>> b = a.reshape((3, 3))
+        >>> torch.norm(a)
+        tensor(7.7460)
+        >>> torch.norm(b)
+        tensor(7.7460)
+        >>> torch.norm(a, float('inf'))
+        tensor(4.)
+        >>> torch.norm(b, float('inf'))
+        tensor([4., 3., 4.])
+        >>> torch.norm(a, 1)
+        tensor(20.)
+        >>> torch.norm(b, 1)
+        tensor(20.)
+        >>> torch.norm(a, -1)
+        tensor(0.)
+        >>> torch.norm(b, -1)
+        tensor(0.)
+        >>> torch.norm(a, 2)
+        tensor(7.7460)
+        >>> torch.norm(b, 2)
+        tensor(7.7460)
+        >>> torch.norm(a, -2)
+        tensor(0.)
+        >>> torch.norm(b, -2)
+        tensor(0.)
+        >>> torch.norm(a, 3)
+        tensor(5.8480)
+        >>> torch.norm(a, -3)
+        tensor(0.)
+        >>> c = torch.tensor([[ 1, 2, 3],[-1, 1, 4]] , dtype= torch.float)
+        >>> torch.norm(c, dim=0)
+        tensor([1.4142, 2.2361, 5.0000])
+        >>> torch.norm(c, dim=1)
+        tensor([3.7417, 4.2426])
+        >>> torch.norm(c, p=1, dim=1)
+        tensor([6., 6.])
+        >>> d = torch.arange(8, dtype= torch.float).reshape(2,2,2)
+        >>> torch.norm(d, dim=(1,2))
+        tensor([ 3.7417, 11.2250])
+        >>> torch.norm(d[0, :, :]), torch.norm(d[1, :, :])
+        (tensor(3.7417), tensor(11.2250))
+    """
+    ndim = input.dim()
+
+    # catch default case
+    if dim is None and out is None:
+        if isinstance(p, int) or (isinstance(p, float) and p != inf and p != -inf):
+            return torch._C._VariableFunctions.norm(input, p)
+        if p == "fro":
+            return torch._C._VariableFunctions.frobenius_norm(input)
+
+    if dim is None:
+        dim = tuple(range(ndim))
+    elif isinstance(dim, list):
+        dim = tuple(dim)
+    elif not isinstance(dim, tuple):
+        try:
+            dim = int(dim)
+        except Exception:
+            raise TypeError(
+                "'dim' must be None, an integer or a tuple/list of integers"
+            )
+        dim = (dim,)
```

Comment:
Why do we need to parse `dim` ourselves? Can't we just always passing it to the C functions, except in `nuc` case we can do some manual checking because it doesn't support `dim` argument.